core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+use std::{fmt::Debug, sync::Arc};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use tokio::{runtime::Handle, sync::watch};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_state::{
+    open_state_keeper_rocksdb, PostgresStorage, ReadStorage, RocksdbStorage, RocksdbStorageBuilder,
+    StateKeeperColumnFamily,
+};
+use zksync_storage::RocksDB;
+use zksync_types::{L1BatchNumber, MiniblockNumber};
+
+/// Factory that can produce a [`ReadStorage`] implementation on demand.
+#[async_trait]
+pub trait ReadStorageFactory: Debug + Send + Sync + 'static {
+    async fn access_storage<'a>(
+        &'a self,
```

Comment:
Nit: The lifetime can be elided here; Rust assumes that the output borrows from `self` by default.